NIFI-12485: Lineage Graph

mcgilman opened the pull request:
NIFI-12485:
- Lineage.
- Context Menu.
- Lineage timeline slider.

rfellows commented:
Will review.

rfellows commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/provenance-event-table.component.ts:
```diff
@@ -77,16 +86,69 @@ export class ProvenanceEventTable implements AfterViewInit {
+                    // ensure this event has an event time
+                    if (minMillis < 0 || minMillis > node.millis) {
+                        minMillis = node.millis;
+                        // minTimestamp = node.timestamp;
```
is the minTimestamp not needed? can we remove this?

mcgilman replied:
Yup!

rfellows commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/lineage/lineage.component.ts:
```diff
@@ -0,0 +1,975 @@
+            },
+            {
+                condition: function (selection: any, self: LineageComponent) {
+                    if (selection.empty()) {
+                        return false;
+                    }
+
+                    const selectionData: any = selection.datum();
+                    return self.supportsExpandCollapse(selectionData);
+                },
+                clazz: 'fa fa-plus-square',
+                text: 'Expand',
+                action: function (selection: any, self: LineageComponent) {
+                    const selectionData: any = selection.datum();
+
+                    // TODO - cluster node id
+                    self.submitLineageQuery.next({
+                        lineageRequestType: 'CHILDREN',
+                        eventId: selectionData.id
+                        // clusterNodeId: clusterNodeId
+                    });
+                }
+            },
+            {
+                condition: function (selection: any, self: LineageComponent) {
+                    if (selection.empty()) {
+                        return false;
+                    }
+
+                    const selectionData: any = selection.datum();
+                    return self.supportsExpandCollapse(selectionData);
+                },
+                clazz: 'fa fa-minus-square',
+                text: 'Collapse',
+                action: function (selection: any, self: LineageComponent) {
+                    const selectionData: any = selection.datum();
+                    self.collapseLineage(selectionData);
+                }
+            }
+        ]
```
I would try to simplify by not requiring the `self` references. You can accomplish this by using arrow functions, then `this` is the LineageComponent.
```suggestion
        menuItems: [
            {
                condition: (selection: any) => selection.empty(),
                clazz: 'fa fa-long-arrow-left',
                text: 'Back to events',
                action: (selection: any) => {
                    this.closeLineage.next();
                }
            },
            {
                condition: (selection: any) => !selection.empty(),
                clazz: 'fa fa-info-circle',
                text: 'View details',
                action: (selection: any) => {
                    const selectionData: any = selection.datum();

                    // TODO cluster node id
                    this.openEventDialog.next({
                        id: selectionData.id
                    });
                }
            },
            {
                condition: (selection: any) => !selection.empty(),
                clazz: 'fa fa-long-arrow-right',
                text: 'Go to component',
                action: (selection: any) =>  {
                    const selectionData: any = selection.datum();
                    this.goToProvenanceEventSource.next({
                        eventId: selectionData.id
                    });
                }
            },
            {
                condition: (selection: any) => {
                    if (selection.empty()) {
                        return false;
                    }

                    const selectionData: any = selection.datum();
                    return this.supportsExpandCollapse(selectionData);
                },
                clazz: 'fa fa-binoculars',
                text: 'Find parents',
                action: (selection: any) => {
                    const selectionData: any = selection.datum();

                    // TODO - cluster node id
                    this.submitLineageQuery.next({
                        lineageRequestType: 'PARENTS',
                        eventId: selectionData.id
                        // clusterNodeId: clusterNodeId
                    });
                }
            },
            {
                condition: (selection: any) => {
                    if (selection.empty()) {
                        return false;
                    }

                    const selectionData: any = selection.datum();
                    return this.supportsExpandCollapse(selectionData);
                },
                clazz: 'fa fa-plus-square',
                text: 'Expand',
                action: (selection: any) => {
                    const selectionData: any = selection.datum();

                    // TODO - cluster node id
                    this.submitLineageQuery.next({
                        lineageRequestType: 'CHILDREN',
                        eventId: selectionData.id
                        // clusterNodeId: clusterNodeId
                    });
                }
            },
            {
                condition: (selection: any) => {
                    if (selection.empty()) {
                        return false;
                    }

                    const selectionData: any = selection.datum();
                    return this.supportsExpandCollapse(selectionData);
                },
                clazz: 'fa fa-minus-square',
                text: 'Collapse',
                action: (selection: any) => {
                    const selectionData: any = selection.datum();
                    this.collapseLineage(selectionData);
                }
            }
        ]
```

rfellows commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/lineage/lineage.component.ts:
```diff
@@ -0,0 +1,975 @@
+                // include if the condition matches
+                if (menuItem.condition) {
+                    const selection: any = d3.select('circle.context');
+                    return menuItem.condition(selection, self);
```
Assuming you take the suggestion about not requiring `self` not needing to be passed in, you can remove it as an argument here.

rfellows commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/lineage/lineage.component.ts:
```diff
@@ -0,0 +1,975 @@
+            menuItemClicked(menuItem: ContextMenuItemDefinition, event: MouseEvent) {
+                if (menuItem.action) {
+                    const selection: any = d3.select('circle.context');
+                    return menuItem.action(selection, self);
```
Assuming you take the suggestion about not requiring `self` not needing to be passed in, you can remove it as an argument here.

rfellows commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/provenance-event-table.component.html:
```diff
@@ -15,142 +15,184 @@
+<div class="provenance-event-table h-full">
+    <div [class.hidden]="showLineage" class="h-full flex flex-col gap-y-2">
+        <div class="flex flex-col">
+            <div class="value font-bold">Displaying {{ filteredCount }} of {{ totalCount }}</div>
```
Consider hiding this unless there is a filter active. It can actually be a bit confusing now that there is pagination. It tells us that 1000 of 1000 are displayed, but with pagination we only ever show 100 so it can be a bit confusing.
Might also consider re-wording it so it is clear that it is referring to the number of items that pass the filter, but not necessarily how many are displayed.

mcgilman replied:
Good call. Will update.

rfellows reviewed (changes requested):
I noticed some rendering issues in the lineage graph sometimes when expanding nodes (in particular FORKs):
<img width="316" alt="Screenshot 2023-12-20 at 4 52 04 PM" src="https://github.com/apache/nifi/assets/713866/90298291-82f8-4f7c-aaac-a1329d025650">

It doesn't happen on all events, but does happen consistently with the certain events. One note, when setting a breakpoint in the lineage.component.ts line 900, and stepping through the expand action, the issue does NOT present itself. So it is probably something with in the timing 😞 

Also, a few things inline...

mcgilman commented:
Thanks for the review @rfellows! I've addressed the items you've mentioned and fixed the rendering issues as well. Please have a look at the latest commit.

rfellows commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/provenance-event-table.component.html:
```diff
@@ -18,7 +18,9 @@
         <div class="flex flex-col">
-            <div class="value font-bold">Displaying {{ filteredCount }} of {{ totalCount }}</div>
+            <div [class.invisible]="!filterApplied" class="value font-bold">
+                Displaying {{ filteredCount }} of {{ totalCount }}
```
This is still not very intuitive IMO. Maybe something like

"Filter matched {{filterCount}} of {{totalCount}}"

Since we still aren't displaying all of the matches if they span more than 1 page.

rfellows commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/provenance/ui/provenance-event-listing/provenance-event-table/lineage/lineage.component.ts:
```diff
@@ -46,11 +46,13 @@ export class LineageComponent implements OnInit {
         if (this.previousEventTimestampThreshold >= 0) {
-            let nodes: any = this.lineageContainerElement.selectAll('g.node');
-            let links: any = this.lineageContainerElement.selectAll('path.link');
+            let nodes: any = this.lineageContainerElement.selectAll('g.node.rendered');
+            let links: any = this.lineageContainerElement.selectAll('path.link.rendered');
```
when showing/hiding these items we should also turn on/of pointer-events. Otherwise, you can still get to the context menus for the hidden items if you mouse over them

rfellows reviewed (changes requested):
Just 2 things and then i think we are good.

rfellows reviewed (approved):
Thanks for the changes, @mcgilman. Looks good.
